### PR TITLE
getStructure

### PR DESCRIPTION
**Pull request by cancerberoSgx**
I think this is ready:
 * scripts/ensureForEachStructureClassImplementsGetStructureMethod.ts to ensure for every structure its corresponding node kind implements getStruture method
 * for each node kind implementing getStructure there is a test (more or less exhaustively) 
 * master merged
 * most of your review comments fixed (still checking one by one to make sure I'm not missing anything)
 * npm i && lint && build && test passing from scratch

### CODE REVIEW

File: src/compiler/decorator/Decorator.ts [hunk 1]
```diff
@@ -274,4 +277,11 @@ export class Decorator extends DecoratorBase<ts.Decorator> {
                 getSiblingFormatting: (parent, sibling) => sibling.getStartLinePos() === thisStartLinePos ? FormattingKind.Space : FormattingKind.Newline
             });
     }
+
+    getStructure(): DecoratorStructure {
+        return {
+            name: this.getName(),
+            arguments: this.getArguments().map(arg => arg.getText())
```
Comment (dsherret) [inline src/compiler/decorator/Decorator.ts]: Careful here. Providing an empty array will make it a decorator factory (ex `@decorator()`). This needs to be undefined when there's no parenthesis (ex. `@decorator`)

File: src/compiler/doc/JSDoc.ts [hunk 2]
```diff
@@ -81,4 +83,11 @@ export class JSDoc extends Node<ts.JSDoc> {
             removeFollowingNewLines: true
         });
     }
+
+    getStructure() {
+        return callBaseGetStructure<JSDocStructure>({}, this, {
+            description: this.getInnerText() // TODO Not sure about this
```
Comment (dsherret) [inline src/compiler/doc/JSDoc.ts]: Yeah, JSDocStructure is lacking at the moment. This is perfect for the time being and you can remove the todo.

File: src/compiler/function/ArrowFunction.ts [hunk 3]
```diff
@@ -3,15 +3,30 @@ import { AsyncableNode, BodiedNode, JSDocableNode, ModifierableNode, SignaturedD
 import { Node } from "../common";
```
Comment (dsherret) [inline src/compiler/function/ArrowFunction.ts]: Can you revert the changes to `ArrowFunction` as part of this pull request (`ArrowFunctionPrinter`, `ArrowFunctionStructure`, etc)? We can add it in a separate pull request.

File: src/compiler/function/FunctionDeclaration.ts [hunk 4]
```diff
@@ -79,4 +80,14 @@ export class FunctionDeclaration extends FunctionDeclarationBase<ts.FunctionDecl
     remove() {
         removeOverloadableStatementedNodeChild(this);
     }
+
+    /**
+     * Gets the structure equivalent to this node
+     */
+    getStructure(): FunctionDeclarationStructure {
+        return callBaseGetStructure<FunctionDeclarationSpecificStructure>(FunctionDeclarationBase.prototype, this, {
+            overloads: undefined // TODO: don't know how to implement this
```
Comment (dsherret) [inline src/compiler/function/FunctionDeclaration.ts]: What about getting the structure from the base and then stripping out any properties that aren't for the overload structure?

File: src/compiler/function/FunctionDeclaration.ts [hunk 5]
```diff
@@ -79,4 +80,14 @@ export class FunctionDeclaration extends FunctionDeclarationBase<ts.FunctionDecl
     remove() {
         removeOverloadableStatementedNodeChild(this);
     }
+
+    /**
+     * Gets the structure equivalent to this node
+     */
+    getStructure(): FunctionDeclarationStructure {
```
Comment (dsherret) [inline src/compiler/function/FunctionDeclaration.ts]: Should return a union with the overload structure.

File: src/compiler/interface/MethodSignature.ts [hunk 6]
```diff
@@ -23,4 +23,11 @@ export class MethodSignature extends MethodSignatureBase<ts.MethodSignature> {
     remove() {
         removeInterfaceMember(this);
     }
+
+    getStructure(): MethodSignatureStructure {
```
Comment (dsherret) [inline src/compiler/interface/MethodSignature.ts]: We need to use `callBaseGetStructure` here.

File: src/compiler/decorator/Decorator.ts [hunk 7]
```diff
@@ -274,4 +277,11 @@ export class Decorator extends DecoratorBase<ts.Decorator> {
                 getSiblingFormatting: (parent, sibling) => sibling.getStartLinePos() === thisStartLinePos ? FormattingKind.Space : FormattingKind.Newline
             });
     }
+
+    getStructure(): DecoratorStructure {
```
Comment (dsherret) [inline src/compiler/decorator/Decorator.ts]: Do the `callBaseGetStructure` here to future proof it.

File: src/compiler/statement/StatementedNode.ts [hunk 8]
```diff
@@ -852,6 +853,17 @@ export function StatementedNode<T extends Constructor<StatementedNodeExtensionTy
             return this;
         }
 
+        getStructure() {
+            return callBaseGetStructure<StatementedNodeStructure>(Base.prototype, this, {
+                classes: [], // TODO: this.getClasses().map(c=>c.getStructure())
```
Comment (dsherret) [inline src/compiler/statement/StatementedNode.ts]: This should work?

File: src/compiler/type/TypeParameterDeclaration.ts [hunk 9]
```diff
@@ -142,6 +144,14 @@ export class TypeParameterDeclaration extends TypeParameterDeclarationBase<ts.Ty
             removeChildren({ children });
         }
     }
+
+    getStructure() {
+        return callBaseGetStructure<TypeParameterDeclarationStructure>({}, this, {
+            constraint: this.getConstraintNode() ? this.getConstraintNode()!.getText() : undefined,
```
Comment (dsherret) [inline src/compiler/type/TypeParameterDeclaration.ts]: Assign this and default to temporary variables.

File: src/structurePrinters/class/MethodDeclarationStructurePrinter.ts [hunk 10]
```diff
@@ -28,7 +28,7 @@ export class MethodDeclarationStructurePrinter extends FactoryStructurePrinter<M
         this.printOverloads(writer, structure.name, getOverloadStructures());
         this.printBase(writer, structure.name, structure);
 
-        if (this.options.isAmbient)
+        if (this.options.isAmbient || structure.isAbstract)
```
Comment (dsherret) [inline src/structurePrinters/class/MethodDeclarationStructurePrinter.ts]: Thanks! Did you add a test for this? Might be worth putting this in a separate pull request though.

File: src/structures/class/PropertyDeclarationStructure.ts [hunk 11]
```diff
@@ -7,3 +7,7 @@ export interface PropertyDeclarationStructure
         JSDocableNodeStructure, ReadonlyableNodeStructure, InitializerExpressionableNodeStructure, DecoratableNodeStructure, AbstractableNodeStructure
 {
 }
+
+export interface PropertyDeclarationSpecificStructure
```
Comment (dsherret) [inline src/structures/class/PropertyDeclarationStructure.ts]: Add this to the extends above to future proof it. Also, move the brace to the same line... I only did that above because there's a long extends clause.

File: src/tests/compiler/class/classDeclarationTests.ts [hunk 12]
```diff
@@ -2,7 +2,8 @@
 import { ClassDeclaration, ConstructorDeclaration, ExpressionWithTypeArguments, GetAccessorDeclaration, MethodDeclaration, ParameterDeclaration,
     PropertyDeclaration, Scope, SetAccessorDeclaration } from "../../../compiler";
 import { ClassDeclarationSpecificStructure, ConstructorDeclarationStructure, GetAccessorDeclarationStructure, MethodDeclarationStructure, PropertyDeclarationStructure,
-    SetAccessorDeclarationStructure } from "../../../structures";
+    SetAccessorDeclarationStructure,
+    ClassDeclarationStructure} from "../../../structures";
```
Comment (dsherret) [inline src/tests/compiler/class/classDeclarationTests.ts]: Put on same line as above and add space before closing brace.

File: src/tests/compiler/class/classDeclarationTests.ts [hunk 13]
```diff
@@ -1347,4 +1348,64 @@ class Child extends Mixin(Base) {}
             doTest("class I {}\n\nclass J {}\n\nclass K {}", 1, "class I {}\n\nclass K {}");
         });
     });
+
+    describe(nameof<ClassDeclaration>(d => d.getStructure), () => {
+        function doTest(code: string, expectedStructure?: ClassDeclarationStructure) {
+            const {descendant, project} = getInfoFromTextWithDescendant<ClassDeclaration>(code, SyntaxKind.ClassDeclaration);
+            const structure = descendant.getStructure();
+            if (expectedStructure) {
+                expect(Object.assign({}, structure, {methods: undefined, properties: undefined, ctors: undefined})).to.contain(
```
Comment (dsherret) [inline src/tests/compiler/class/classDeclarationTests.ts]: We should probalby do exact comparisons between the structures in the tests.

File: src/tests/compiler/class/classDeclarationTests.ts [hunk 14]
```diff
@@ -1347,4 +1348,64 @@ class Child extends Mixin(Base) {}
...
+                    Object.assign({}, expectedStructure, {methods: undefined, properties: undefined, ctors: undefined}));
+            }
+            const sourceFile2 = project.createSourceFile("__tmp_sourceFile1.ts", "", {overwrite: true});
```
Comment (dsherret) [inline src/tests/compiler/class/classDeclarationTests.ts]: You can just add to the same file as above and then compare the texts of the two classes.

File: src/tests/compiler/class/classDeclarationTests.ts [hunk 15]
```diff
@@ -1347,4 +1348,64 @@ class Child extends Mixin(Base) {}
...
+            const generatedClassDecl = sourceFile2.addClass(structure);
+            expect(generatedClassDecl.getText().replace(/\s+/gm, "")).equals(descendant.getText().replace(/\s+/gm, ""));
```
Comment (dsherret) [inline src/tests/compiler/class/classDeclarationTests.ts]: Good idea to do this text comparison. Do we need to remove the spaces though?

File: src/tests/compiler/class/classDeclarationTests.ts [hunk 16]
```diff
@@ -1347,4 +1348,64 @@ class Child extends Mixin(Base) {}
             doTest("class I {}\n\nclass J {}\n\nclass K {}", 1, "class I {}\n\nclass K {}");
         });
     });
+
+    describe(nameof<ClassDeclaration>(d => d.getStructure), () => {
+        function doTest(code: string, expectedStructure?: ClassDeclarationStructure) {
```
Comment (dsherret) [inline src/tests/compiler/class/classDeclarationTests.ts]: Use `MakeRequired<ClassDeclaration>` here to ensure we test all the properties of `ClassDeclarationStructure`.

File: src/tests/compiler/class/methodDeclarationTests.ts [hunk 17]
```diff
@@ -109,6 +111,23 @@ describe(nameof(MethodDeclaration), () => {
...
 
+    describe(nameof<MethodDeclaration>(m => m.isOverload), () => {
+        function doTest(startingCode: string, methodName: string, expected: boolean) {
+            const {firstChild, sourceFile} = getInfoFromText<ClassDeclaration>(startingCode);
+            const method = firstChild.getInstanceMethods().find(m => m.getName() === methodName);
+            expect(method).to.not.be.undefined;
+            expect(method!.isOverload()).to.be.equals(expected);
+        }
+
+        xit("should be false for a non overloaded abstract method ?", () => {
```
Comment (dsherret) [inline src/tests/compiler/class/methodDeclarationTests.ts]: xit? This compiles?

Comment (dsherret) [inline src/tests/compiler/class/methodDeclarationTests.ts]: Remove question mark at end of sentence

File: src/tests/compiler/class/methodDeclarationTests.ts [hunk 18]
```diff
@@ -109,6 +111,23 @@ describe(nameof(MethodDeclaration), () => {
...
+            doTest(`
```
Comment (dsherret) [inline src/tests/compiler/class/methodDeclarationTests.ts]: We'll have to add tests for other scenarios for this. Perhaps as part of a different pull request though.

File: src/tests/compiler/class/methodDeclarationTests.ts [hunk 19]
```diff
@@ -202,4 +221,36 @@ describe(nameof(MethodDeclaration), () => {
...
+            (expectedStructure.parameters || []).forEach((expectedParameter, i) => {
+                expect(structure.parameters![i], `${i}th parameter of method ${structure.name}`).to.contain(expectedParameter);
+            });
+            // and also recreate the AST using structure and compare getText() of generated code with original node's
+            const aux = sourceFile.addStatements("class __AuxClass{}");
+            const decl = aux[0];
+            if (!TypeGuards.isClassDeclaration(decl)) {
+                return expect("!TypeGuards.isClassDeclaration(decl)").to.be.undefined;
+            }
+            expect(decl.getMethods().length).to.eq(0);
```
Comment (dsherret) [inline src/tests/compiler/class/methodDeclarationTests.ts]: This is really specific to the one test.

File: src/tests/compiler/class/methodDeclarationTests.ts [hunk 20]
```diff
@@ -202,4 +221,36 @@ describe(nameof(MethodDeclaration), () => {
...
+            const structure = descendant.getStructure();
+            expect(Object.assign({}, structure, {parameters: undefined})).to.contain(Object.assign({}, expectedStructure, {parameters: undefined}));
+            expect(expectedStructure.parameters && expectedStructure.parameters.length).to.equals(structure.parameters && structure.parameters.length);
+            (expectedStructure.parameters || []).forEach((expectedParameter, i) => {
+                expect(structure.parameters![i], `${i}th parameter of method ${structure.name}`).to.contain(expectedParameter);
+            });
+            // and also recreate the AST using structure and compare getText() of generated code with original node's
+            const aux = sourceFile.addStatements("class __AuxClass{}");
+            const decl = aux[0];
+            if (!TypeGuards.isClassDeclaration(decl)) {
```
Comment (dsherret) [inline src/tests/compiler/class/methodDeclarationTests.ts]: Don't worry about adding this. It's obvious from above that it's a class so just assert it.

File: src/tests/compiler/class/methodDeclarationTests.ts [hunk 21]
```diff
@@ -202,4 +221,36 @@ describe(nameof(MethodDeclaration), () => {
             });
         });
     });
+
+    describe(nameof<MethodDeclaration>(d => d.getStructure), () => {
+        function doTest(code: string, expectedStructure: MethodDeclarationStructure) {
```
Comment (dsherret) [inline src/tests/compiler/class/methodDeclarationTests.ts]: Try using `MakeRequired<MethodDeclarationStructure>`

File: src/tests/compiler/function/functionDeclarationTests.ts [hunk 22]
```diff
@@ -133,4 +134,45 @@ describe(nameof(FunctionDeclaration), () => {
...
+
+    describe(nameof<FunctionDeclaration>(d => d.getStructure), () => {
+        // this is comparing original Node.getText() with a the getText() of a new Node filled with generated structure.
+        // Compared strings have no spaces so comparision doesn't fail.
+        function doTest(text: string) {
+            const {firstChild, sourceFile} = getInfoFromText<FunctionDeclaration>(text);
+            const structure = firstChild.getStructure();
+            const aux = sourceFile.addStatements("function aux(){}");
+            const decl = aux[0];
+            if (!TypeGuards.isFunctionDeclaration(decl)) {
```
Comment (dsherret) [inline src/tests/compiler/function/functionDeclarationTests.ts]: Can remove this too. Just do an assertion.

File: src/tests/compiler/function/parameterDeclarationTests.ts [hunk 23]
```diff
@@ -129,4 +130,25 @@ describe(nameof(ParameterDeclaration), () => {
             doTest("function identifier(param, param2, param3) {}", "param2", "function identifier(param, param3) {}");
         });
     });
+
+    describe(nameof<ParameterDeclaration>(d => d.getStructure), () => {
+        function doTest(code: string, expectedStructure: ParameterDeclarationStructure) {
```
Comment (dsherret) [inline src/tests/compiler/function/parameterDeclarationTests.ts]: `MakeRequired<ParameterDeclarationStructure>`

File: src/tests/compiler/function/parameterDeclarationTests.ts [hunk 24]
```diff
@@ -129,4 +130,25 @@ describe(nameof(ParameterDeclaration), () => {
...
+            const {descendant} = getInfoFromTextWithDescendant<ParameterDeclaration>(code, SyntaxKind.Parameter);
+            const structure = descendant.getStructure();
+            expect(structure).to.contain(expectedStructure);
```
Comment (dsherret) [inline src/tests/compiler/function/parameterDeclarationTests.ts]: We should do an exact comparison in case anything changes. `expect(structure).to.deep.equal(expectedStructure)`.

File: src/compiler/statement/StatementedNode.ts (hunk 8 above)
Comment (dsherret) [reply src/compiler/statement/StatementedNode.ts]: Ah, I just realized why this is hard to do. Ok, for this just leave everything undefined. It will rely on the body text only for these kind of nodes (so source files and namespaces will only have their body text too when doing `.getStructure()`). Too difficult to handle any other way.

File: src/tests/compiler/class/classDeclarationTests.ts (hunk 15 above)
Comment (dsherret) [reply src/tests/compiler/class/classDeclarationTests.ts]: I was just thinking about this more. Could we remove the add with text comparisons? That way we're testing the structures directly which would catch any issues with it.

File: src/tests/compiler/function/parameterDeclarationTests.ts (hunk 24 above)
Comment (cancerberoSgx) [reply src/tests/compiler/function/parameterDeclarationTests.ts]: sure, BTW right now for quickly testing I'm not doing what I've should, this is asserting on the structure object itself instead of re-render a new node asserting against the node.getText() - but at least it gives me some confidence I'm doing things OK and that it works ll the way down to text.

**Comment by cancerberoSgx** [comment]
So, I made all the changes you asked David, with the exception of ` if (!TypeGuards.isClassDeclaration(decl))`, that's my way of type-guard instead of using orThrows and also this one : 

> Hmmm, could we always use the same base (ConstructorDeclarationBase.prototype), but instead strip the unused properties when this.isOverload() is true? Then we can catch changes to the base that shouldn't be in there in a unit test.

**Comment by cancerberoSgx** [comment]
Regarding: 

> By the way, if at any time you want to merge this I can merge it into a branch and then I'll pick up the work to finish it off.

I would really like to have this feature in the library since I'm generating a log of code in my projects and basically working with my handcrafted "getStructure" methods. If you think it's more practical that you finish off the work in your own branch then go ahead (I feel you wasted more time in th elast review than if you implemented by yourself.) But to be sincere I would like to keep working on this and give support for getStructure to the rest of the nodes. But primariy, as I said before I would really need this integrated in master ASAP. 

Thanks!

**Comment by dsherret** [comment]
@cancerberoSgx I don't view it as wasting time because you taking on this work saves me a lot of time and plus it's another person who learns the inner workings of the library, which is always a plus. I just provided that as an option in case you didn't want to spend the time changing all my comments.

I'm going to be moving over the next week (lots of time spent packing) and attending some family events so I won't have time to work on this for the next while. Please continue working on it 😄

**Comment by cancerberoSgx** [comment]
Nice thanks! BTW I also want to take this opportunity to learn about the inners of the library. I must say it was kind of intuitive once you point me how to workaround that mysterious build error :) Also I must say you are a very careful developer (compared to me) :P What do you think is missing in order to merge this work ? SHould I keep implementing getStructure for the rest of Node kinds ? As I said before I'm kind o fan urgency of having this feature and I woulnd't want to use my own branch... Please advice. 

BTW I'm using the library to build TLS plugins for code fixes and refactors common in other typed languages / ides like Java/eclipse.  See https://github.com/cancerberoSgx/typescript-plugins-of-mine/tree/master/typescript-plugin-proactive-code-fixes - there are some .gifs at the bottom.  I bet you didn't imagine tsa being used like this. It saved me a lot of time although some of the refactors could take more than one second to apply ... Just experimenting and learning - let's see if I reach a production - ready product some day. 

THanks!

**Comment by dsherret** [comment]
For it to be merged into master this would have to be complete. So for the definition of complete I would say:

1. For every currently existing `Structure` interface there should exist a corresponding `.getStructure()` method on the node.
2. For every `.getStructure()` there should exist a test that checks the return value. <-- we don't need to go over the top and test every combination, but rather just ensure things generally seem to be working (maybe ballpark 1-3 tests per `.getStructure()` method, but obviously use discretion in the scenario)

If you urgently need this feature, then I recommend publishing your fork to npm and using that package in what you need `.getStructure()` for.

RE the plugins: very cool! For performance, have you looked into why it's taking so long? Should be pretty quick if you're just looking at the AST and not any symbols or types, but yes, it could be a bit slow otherwise. By the way, see the performance tip I recently added here: https://dsherret.github.io/ts-simple-ast/manipulation/performance -- not sure if it would help though.

**Comment by cancerberoSgx** [comment]
Re plugins performance, I'm creating a new project every time a refactor is applied. Probably this can be improved - but the most slow thing is typeChecker.getTypeAtPosition or type inferring in general - no matter if the project is reused or not. Sincerily didn't put much focus in performance right now but I will, thanks!

**Comment by dsherret** [comment]
@cancerberoSgx thanks for still working on this! Let me know when you'd like me to take a look again.

**Comment by dsherret** [comment]
I can do a commit to fix the conflicts. I'll probably do it next weekend though.

**Comment by cancerberoSgx** [comment]
I'm looking at the changes and I need time to understand how to merge these: NamedNode, BindingNamedNode correctly using NamedNodeBase. I don't have much time this week probably next one I can do it - would be ideal if you help / do it - thanks! ;P

**Comment by dsherret** [comment]
Hey @cancerberoSgx, thanks a lot once again for this! It's easier for me to do the remaining work so I'll be working on this over the next while and hopefully we can get this merged this month.

File: src/compiler/class/ClassDeclaration.ts [hunk 25]
```diff
@@ -906,6 +907,23 @@ export class ClassDeclaration extends ClassDeclarationBase<ts.ClassDeclaration>
...
 
+    /**
+     * Gets the structure equivalent to this node.
+     */
+    getStructure(): ClassDeclarationStructure {
+        const getExtends = this.getExtends();
+        return callBaseGetStructure<ClassDeclarationSpecificStructure>(ClassDeclarationBase.prototype, this, {
+            ctors: this.getConstructors().filter(ctor => !ctor.isOverload())
+                .map(ctor => ctor.getStructure()),
+            methods: this.getMethods().filter(method => !method.isOverload() || method.isAbstract())
```
Comment (dsherret) [inline src/compiler/class/ClassDeclaration.ts]: Heh, this is a bug in the library. `isOverload()` should return false for abstract members.

File: src/compiler/class/PropertyDeclaration.ts [hunk 26]
```diff
@@ -36,4 +37,12 @@ export class PropertyDeclaration extends PropertyDeclarationBase<ts.PropertyDecl
                 throw new errors.NotImplementedError(`Not implemented parent syntax kind: ${parent.getKindName()}`);
         }
     }
+
+    /**
+     * Gets the structure equivalent to this node
+     */
+    getStructure(): PropertyDeclarationStructure {
```
Comment (dsherret) [inline src/compiler/class/PropertyDeclaration.ts]: Needs tests. (I'll add them later. Just reminding myself)

File: src/compiler/decorator/Decorator.ts [hunk 27]
```diff
@@ -2,9 +2,12 @@ import * as errors from "../../errors";
 import { FormattingKind, insertIntoParentTextRange, removeChildren, removeChildrenWithFormattingFromCollapsibleSyntaxList } from "../../manipulation";
```
Comment (dsherret) [inline src/compiler/decorator/Decorator.ts]: Needs tests.

File: src/compiler/doc/JSDoc.ts [hunk 28]
```diff
@@ -5,6 +5,8 @@ import { ts } from "../../typescript";
 import { getTextFromStringOrWriter } from "../../utils";
```
Comment (dsherret) [inline src/compiler/doc/JSDoc.ts]: Needs tests.

File: src/compiler/expression/object/SpreadAssignment.ts [hunk 29]
```diff
@@ -3,6 +3,8 @@ import { Node } from "../../common/Node";
 import { ExpressionedNode } from "../expressioned";
```
Comment (dsherret) [inline src/compiler/expression/object/SpreadAssignment.ts]: Needs tests.

File: src/compiler/expression/object/ShorthandPropertyAssignment.ts [hunk 30]
```diff
@@ -94,4 +96,12 @@ export class ShorthandPropertyAssignment extends ShorthandPropertyAssignmentBase
...
         removeCommaSeparatedChild(this);
     }
+
+    /**
+     * Gets the structure equivalent to this node.
+     */
+    getStructure(): ShorthandPropertyAssignmentStructure {
+        return callBaseGetStructure<ShorthandPropertyAssignmentSpecificStructure>(ShorthandPropertyAssignmentBase.prototype, this, {
+        }) as any as ShorthandPropertyAssignmentStructure;
+    }
```
Comment (dsherret) [inline src/compiler/expression/object/ShorthandPropertyAssignment.ts]: Needs tests.

File: src/compiler/file/ImportSpecifier.ts [hunk 31]
```diff
@@ -86,4 +88,15 @@ export class ImportSpecifier extends Node<ts.ImportSpecifier> {
...
+    /**
+     * Gets the structure equivalent to this node.
+     */
+    getStructure() {
+        const alias = this.getAliasNode();
+        return callBaseGetStructure<ImportSpecifierStructure>(Node.prototype, this, {
+            name: this.getName(),
+            alias: alias ? alias.getText() : undefined
+        });
+    }
```
Comment (dsherret) [inline src/compiler/file/ImportSpecifier.ts]: Needs tests.

File: src/compiler/interface/CallSignatureDeclaration.ts [hunk 32]
```diff
@@ -23,4 +24,12 @@ export class CallSignatureDeclaration extends CallSignatureDeclarationBase<ts.Ca
...
         removeInterfaceMember(this);
     }
+
+    /**
+     * Gets the structure equivalent to this node.
+     */
+    getStructure(): CallSignatureDeclarationStructure {
+        return callBaseGetStructure<CallSignatureDeclarationSpecificStructure>(CallSignatureDeclarationBase.prototype, this, {
+        });
+    }
```
Comment (dsherret) [inline src/compiler/interface/CallSignatureDeclaration.ts]: Needs tests.

File: src/compiler/type/TypeParameterDeclaration.ts [hunk 33]
```diff
@@ -142,6 +144,14 @@ export class TypeParameterDeclaration extends TypeParameterDeclarationBase<ts.Ty
...
+            name: this.getName(),
+            default: this.getDefault() ? this.getDefault()!.getText() : undefined
+        });
+    }
```
Comment (dsherret) [inline src/compiler/type/TypeParameterDeclaration.ts]: Needs tests.

File: src/tests/compiler/class/constructorDeclarationTests.ts [hunk 34]
```diff
@@ -143,4 +143,21 @@ describe(nameof(ConstructorDeclaration), () => {
             doTest("class identifier {\n    constructor() {}\n}", structure, "class identifier {\n    constructor(param);\n    constructor() {}\n}");
         });
     });
+
+    describe(nameof<ConstructorDeclaration>(n => n.getStructure), () => {
+        it("should get structure", () => {
+            const { firstChild } = getInfoFromText<ClassDeclaration>(`
+            class A<T> {
+                /** something in the way she moves */
```
Comment (dsherret) [inline src/tests/compiler/class/constructorDeclarationTests.ts]: I also enjoy this song :)

File: src/tests/compiler/class/methodDeclarationTests.ts (hunk 17 above)
Comment (dsherret) [inline src/tests/compiler/class/methodDeclarationTests.ts]: Ah I see. Yes, this was a bug. I fixed this in master based on a previous part in the review.

File: src/tests/compiler/enum/enumMemberTests.ts [hunk 35]
```diff
@@ -101,4 +101,29 @@ describe(nameof(EnumMember), () => {
...
+        });
+
+        it("should respect names with spaces initializers and docs", () => {
+            doTest(`
+            enum b {
+                /** Use when opinionated */
+                'a very peculiar name' = 3.14
+            }`,
+                {
+                    name: "\'a very peculiar name\'", // TODO: Should the quotes be part of the name ? (I don't think so)
```
Comment (dsherret) [inline src/tests/compiler/enum/enumMemberTests.ts]: Yes, the quotes need to be part of the name because otherwise it would represent a name without quotes.

File: src/tests/compiler/interface/interfaceDeclarationTests.ts [hunk 36]
```diff
@@ -26,14 +27,20 @@ describe(nameof(InterfaceDeclaration), () => {
             doTest("interface Base1 {} interface Base2 {} interface Child extends Base1, Base2 {}", "Child", ["Base1", "Base2"]);
         });
 
+        xit("should get the base when there are multiple with type parameters", () => { // TODO: issue !
+            const { sourceFile } = getInfoFromText(`interface Base1 {} interface A {} interface Child extends Base1<A>`);
+            const types = sourceFile.getInterfaceOrThrow("Child").getBaseTypes();
+            expect(types).not.to.be.empty;
+        });
```
Comment (dsherret) [inline src/tests/compiler/interface/interfaceDeclarationTests.ts]: Not an issue because there's a compile error here. Needs to be `interface Base1<T>`.

File: src/compiler/interface/CallSignatureDeclaration.ts (hunk 32 above)
Comment (cancerberoSgx) [reply src/compiler/interface/CallSignatureDeclaration.ts]: do you need me to add these missing tests ?

Comment (dsherret) [reply src/compiler/interface/CallSignatureDeclaration.ts]: @cancerberoSgx sure, that would be very helpful. Thanks! Just copy what's done in [classDeclarationTests.ts](https://github.com/dsherret/ts-simple-ast/blob/162276a6aed164bc1191a04867b8a451e5422b51/src/tests/compiler/class/classDeclarationTests.ts#L1355). Basically, one test for checking the structure when it doesn't have anything and another for when it has everything.

* Don't bother testing the structures of any children that can also get a structure. Just strip that child structure down to the bare minimum (ex. when testing the class just look at the method names in the class). Then write tests separately for the method's structure.
* Only test getting the structure and not filling or anything else. The tests should be focused on what they're testing.

When finished writing tests for a file please commit and push it right away. That way if we're both working in parallel we won't overlap work and if we do it's not that big of a deal because only one file would have changed.

Thanks a lot once again! 😄

**Comment by dsherret** [comment]
This is merged now, but will be released in [v15](https://github.com/dsherret/ts-simple-ast/milestone/6). Thanks for the help @cancerberoSgx! This was so much work.

**Comment by dsherret** [comment]
Hey @cancerberoSgx, this is released now and seems to work well (in 15.0.1). The `.fill` method is deprecated now though. Use `.set` instead or the `.addX` methods.
